source/ada/lsp-unit_providers.adb

```diff
@@ -0,0 +1,207 @@
+------------------------------------------------------------------------------
+--                         Language Server Protocol                         --
+--                                                                          --
+--                       Copyright (C) 2021, AdaCore                        --
+--                                                                          --
+-- This is free software;  you can redistribute it  and/or modify it  under --
+-- terms of the  GNU General Public License as published  by the Free Soft- --
+-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
+-- sion.  This software is distributed in the hope  that it will be useful, --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public --
+-- License for  more details.  You should have  received  a copy of the GNU --
+-- General  Public  License  distributed  with  this  software;   see  file --
+-- COPYING3.  If not, go to http://www.gnu.org/licenses for a complete copy --
+-- of the license.                                                          --
+------------------------------------------------------------------------------
+
+with Ada.Strings.Unbounded; use Ada.Strings.Unbounded;
+with GNATCOLL.VFS;          use GNATCOLL.VFS;
+
+with Libadalang.Unit_Files;
+with Langkit_Support.Text; use Langkit_Support.Text;
+
+with LSP.Preprocessor; use LSP.Preprocessor;
+
+package body LSP.Unit_Providers is
+
+   package US renames Ada.Strings.Unbounded;
+   use type US.Unbounded_String;
+
+   type Project_Unit_Provider is new LAL.Unit_Provider_Interface with record
+      Tree             : Prj.Project_Tree_Access;
+      Projects         : Prj.Project_Array_Access;
```

Comment:
Currently this provider always uses a single project. Why not replace the array with a single project?